src/jdk.incubator.concurrent/share/classes/jdk/incubator/concurrent/StructuredTaskScope.java

```diff
@@ -0,0 +1,1175 @@
+import java.security.AccessController;
+import java.security.PrivilegedAction;
+import java.time.Duration;
+import java.time.Instant;
+import java.util.Comparator;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.Set;
+import java.util.concurrent.Callable;
+import java.util.concurrent.CancellationException;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.Future;
+import java.util.concurrent.FutureTask;
+import java.util.concurrent.RejectedExecutionException;
+import java.util.concurrent.ThreadFactory;
+import java.util.concurrent.TimeoutException;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.locks.ReentrantLock;
+import java.util.function.Function;
+import jdk.internal.misc.PreviewFeatures;
+import jdk.internal.misc.ThreadFlock;
+
+/**
+ * A basic API for <em>structured concurrency</em>. StructuredTaskScope supports cases
+ * where a task splits into several concurrent sub-tasks, to be executed in their own
+ * threads, and where the sub-tasks must complete before the main task continues. A
+ * StructuredTaskScope can be used to ensure that the lifetime of a concurrent operation
+ * is confined by a <em>syntax block</em>, just like that of a sequential operation in
+ * structured programming.
+ *
+ * <h2>Basic usage</h2>
+ *
+ * A StructuredTaskScope is created with one of its public constructors. It defines
+ * the {@link #fork(Callable) fork} method to start a thread to execute a task, the {@link
+ * #join() join} method to wait for all threads to finish, and the {@link #close() close}
+ * method to close the task scope. The API is intended to be used with the {@code
+ * try-with-resources} construct. The intention is that code in the <em>block</em> uses
+ * the {@code fork} method to fork threads to execute the sub-tasks, wait for the threads
+ * to finish with the {@code join} method, and then <em>process the results</em>.
+ * Processing of results may include handling or re-throwing of exceptions.
+ * {@snippet lang=java :
+ *     try (var scope = new StructuredTaskScope<Object>()) {
+ *
+ *         Future<Integer> future1 = scope.fork(task1);   // @highlight substring="fork"
+ *         Future<String> future2 = scope.fork(task2);    // @highlight substring="fork"
+ *
+ *         scope.join();                                  // @highlight substring="join"
+ *
+ *         ... process results/exceptions ...
+ *
+ *     } // close                                         // @highlight substring="close"
+ * }
+ * To ensure correct usage, the {@code join} and {@code close} methods may only be invoked
+ * by the <em>owner</em> (the thread that opened/created the StructuredTaskScope), and the
+ * {@code close} method throws an exception after closing if the owner did not invoke the
+ * {@code join} method after forking.
+ *
+ * <p> StructuredTaskScope defines the {@link #shutdown() shutdown} method to shut down a
```

Comment:
This sentence, because of the place where it appears, is a bit confusing. So far we only know about the fact that a scope has an owner thread. So it seems odd that shutdown could be called _while_ the owner thread is waiting on a `join`. Of course, then you read what's next, and you discover that: (a) shutdown might be called by a custom scope subclass and that (b) shutdown is confined to the threads contained in this task scope - but this definition is only given much later.

Reply:
I see your point. The intention is to introduce all the public methods before introducing the subclasses or policies. I think I can adjust this sentence to make it clear that a subtask may call shutdown while the owner is waiting in join.